### PR TITLE
[JENKINS-57896] Launch Kafka on K8s function

### PR DESCRIPTION
**Pull request by longngn**
Sorry, I messed up last PR's git history by rebasing. This one would be cleaner
Also reply to @oleg-nenashev :

> It will be extremely hard to detach a strict dependency on K8s plugin after such change. Maybe you want to consider extension points and detached configuration controls

Actually we don't have dependency on k8s plugin anymore. We'll implement our own connector class and Cloud API.

### CODE REVIEW

File: plugin/src/main/java/io/jenkins/plugins/remotingkafka/GlobalKafkaConfiguration.java [hunk 1]
```diff
@@ -268,15 +329,93 @@ public FormValidation doStartKafkaOnKubernetes(
             @QueryParameter("kubernetesSkipTlsVerify") boolean skipTlsVerify,
             @QueryParameter("kubernetesNamespace") String namespace
     ) {
-        // TODO: Use K8s client to launch Zookeeper and Kafka pods
-        return FormValidation.ok("Success");
+        Jenkins.get().checkPermission(Jenkins.ADMINISTER);
+
+        Class clazz = GlobalKafkaConfiguration.class;
+        try (InputStream zookeeperFile = clazz.getResourceAsStream("kubernetes/zookeeper.yaml");
```
Comment (afalko) [inline plugin/src/main/java/io/jenkins/plugins/remotingkafka/GlobalKafkaConfiguration.java]: Best to do `Paths.get("kubernetes", "zookeeper.yaml")`, etc for the others

File: plugin/src/main/java/io/jenkins/plugins/remotingkafka/GlobalKafkaConfiguration.java [hunk 2]
```diff
@@ -268,15 +329,93 @@ public FormValidation doStartKafkaOnKubernetes(
...
+                throw new RemotingKafkaConfigurationException("Couldn't find StatefulSet named kafka in YAML configuration");
+            Container kafkaContainer = kafkaStatefulSet.
+                    getSpec().
+                    getTemplate().
+                    getSpec().
+                    getContainers().
+                    stream().
+                    filter(con -> con.getName().equals("kafka")).
+                    findFirst().
+                    orElse(null);
```
Comment (afalko) [inline plugin/src/main/java/io/jenkins/plugins/remotingkafka/GlobalKafkaConfiguration.java]: Why not throw in the `orElse` and remove the null check below?

File: plugin/src/main/java/io/jenkins/plugins/remotingkafka/GlobalKafkaConfiguration.java [hunk 3]
```diff
@@ -268,15 +329,93 @@ public FormValidation doStartKafkaOnKubernetes(
...
+
+            // Probe for Kafka readiness
+            while (true) {
+                try {
+                    testConnection(serverIp, kafkaPort);
+                    break;
+                } catch (IOException e) {
+                    LOGGER.info(String.format("Waiting for Kafka connection at %s:%s", serverIp, kafkaPort));
+                }
+                Thread.sleep(2000);
```
Comment (afalko) [inline plugin/src/main/java/io/jenkins/plugins/remotingkafka/GlobalKafkaConfiguration.java]: Use `TimeUnit`...maybe also make the sleep a global; 1 second is a more canonical default

File: plugin/src/main/java/io/jenkins/plugins/remotingkafka/GlobalKafkaConfiguration.java [hunk 4]
```diff
@@ -268,15 +329,93 @@ public FormValidation doStartKafkaOnKubernetes(
...
+            }
+            LOGGER.info("Zookeeper and Kafka started");
+
+            // Set Zookeeper and Broker URL
+            GlobalKafkaConfiguration.get().setZookeeperURL(serverIp + ":" + zookeeperPort);
```
Comment (afalko) [inline plugin/src/main/java/io/jenkins/plugins/remotingkafka/GlobalKafkaConfiguration.java]: I suggest URIBuider: https://hc.apache.org/httpcomponents-client-ga/httpclient/apidocs/org/apache/http/client/utils/URIBuilder.html

File: plugin/src/main/java/io/jenkins/plugins/remotingkafka/KubernetesUtils.java [hunk 5]
```diff
@@ -0,0 +1,14 @@
+package io.jenkins.plugins.remotingkafka;
+
+import io.fabric8.kubernetes.api.model.ServicePort;
+import io.fabric8.kubernetes.client.KubernetesClient;
+
+import java.util.List;
+
+public class KubernetesUtils {
```
Comment (afalko) [inline plugin/src/main/java/io/jenkins/plugins/remotingkafka/KubernetesUtils.java]: It is best not to name things "Utils"...maybe call this `KubernetesQuery`?

File: plugin/src/main/resources/io/jenkins/plugins/remotingkafka/kubernetes/zookeeper.yaml [hunk 6]
```diff
@@ -0,0 +1,80 @@
...
+              mountPath: /zookeeper
+  volumeClaimTemplates:
+    - metadata:
+        name: zookeeper-data
+      spec:
+        accessModes: [ "ReadWriteOnce" ]
+        storageClassName: standard
+        resources:
+          requests:
+            storage: 1Gi
```
Comment (afalko) [inline plugin/src/main/resources/io/jenkins/plugins/remotingkafka/kubernetes/zookeeper.yaml]: I think you can make this much smaller. 100Mi is what I'd set this to

File: plugin/src/main/resources/io/jenkins/plugins/remotingkafka/kubernetes/kafka-statefulset.yaml [hunk 7]
```diff
@@ -0,0 +1,66 @@
...
+    type: RollingUpdate
+  podManagementPolicy: Parallel
+  template:
+    metadata:
+      labels:
+        app: kafka
+    spec:
+      containers:
+        - name: kafka
+          image: wurstmeister/kafka:2.12-2.2.0
```
Comment (afalko) [inline plugin/src/main/resources/io/jenkins/plugins/remotingkafka/kubernetes/kafka-statefulset.yaml]: 2.2.1 is the latest version

File: plugin/src/main/java/io/jenkins/plugins/remotingkafka/KubernetesUtils.java (hunk 5 above)
Comment (longngn) [reply plugin/src/main/java/io/jenkins/plugins/remotingkafka/KubernetesUtils.java]: @afalko Why should we not name things "Utils"?

Comment (afalko) [reply plugin/src/main/java/io/jenkins/plugins/remotingkafka/KubernetesUtils.java]: "Utils", "Helper" aren't good at describing what the class is doing and its purpose. This was a tough habit for me to kick as well.

Here's a better elucidated argument: https://gsferreira.com/archive/2017/09/please-don-t-name-it-utils/

Comment (pvtuan10) [reply plugin/src/main/java/io/jenkins/plugins/remotingkafka/KubernetesUtils.java]: @afalko is this a preference or a general programming practice? TBH I am still seeing in a lot of places that people use helpers, utils, ....

**Comment by longngn** [comment]
@afalko @pvtuan10 I pushed a new commit. I decided not to use ConfigMap because ConfigMap can only be used in container template so the internal and external ports would still be scattered in service YAML and readinessProbe. IMO it's not worth an extra layer of complexity.
Also, I tried using Kafka 2.2.1 and it had a problem where it constantly tried to connect to the internal host for several minutes when starting up, so I reverted back to 2.2.0

**Comment by longngn** [comment]
@afalko Is successThreshold really necessary for standalone Kafka? I found it significantly increase start up time (took at least `periodSeconds * successThreshold` seconds)

**Comment by afalko** [comment]
Are we setting period seconds too high? We can revisit this in subsequent PRs, including the Kafka version upgrade.

**Comment by longngn** [comment]
@afalko 10 is the default periodSeconds. Yeah I think we can create a JIRA ticket for those issues

File: plugin/src/main/java/io/jenkins/plugins/remotingkafka/GlobalKafkaConfiguration.java [hunk 8]
```diff
@@ -268,15 +328,89 @@ public FormValidation doStartKafkaOnKubernetes(
...
+
+            // Probe for Kafka readiness
+            while (true) {
+                try {
+                    testConnection(serverIp, kafkaPort);
+                    break;
+                } catch (IOException e) {
+                    LOGGER.info(String.format("Waiting for Kafka connection at %s:%s", serverIp, kafkaPort));
+                }
+                Thread.sleep(TimeUnit.SECONDS.toMillis(1));
```
Comment (afalko) [inline plugin/src/main/java/io/jenkins/plugins/remotingkafka/GlobalKafkaConfiguration.java]: Small trick: TimeUnit.SECONDS.sleep(1)

**Review by afalko — APPROVED** [review]

File: plugin/src/main/java/io/jenkins/plugins/remotingkafka/KubernetesUtils.java (hunk 5 above)
Comment (afalko) [reply plugin/src/main/java/io/jenkins/plugins/remotingkafka/KubernetesUtils.java]: It is a general programming practice. Using "helpers", "utils" is considered lazy - at least according to a lot of the people I've worked with.

**Review by afalko — APPROVED** [review]

**Review by pvtuan10 — APPROVED** [review]

**Review by Supun94 — APPROVED** [review]

**Comment by longngn** [comment]
@afalko I have a new idea. We can use [https://kubernetes.io/docs/concepts/workloads/pods/init-containers/](init container) to wait for Zookeeper servers before creating Kafka container. That way, we don't have to worry about Kafka container's readinessProbe taking too long or too short.

**Comment by afalko** [comment]
@longngn : That sounds promising :). If it works better than what you current have, then :+1:

**Pull request merged**